src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+     */
+    if (!page.test_flag(page::flags::queued)) {
+        log.debug("Marking inflight I/O to be requeued");
+        page.set_flag(page::flags::queued);
+    }
+}
+
+void io_queue::enqueue_pending(page& page) noexcept {
+    pending_.push_back(page);
+    page.set_flag(page::flags::queued);
+    cond_.signal();
+}
+
+seastar::future<> io_queue::dispatch() noexcept {
+    while (true) {
+        co_await cond_.wait([this] { return !pending_.empty() || stop_; });
+
+        if (stop_) {
+            co_return;
+        }
+
+        auto units = co_await seastar::get_units(ops_, 1);
+
+        if (file_ == nullptr) {
+            continue;
+        }
+
+        auto& page = pending_.front();
+        pending_.pop_front();
+        assert(page.test_flag(page::flags::queued));
+        page.clear_flag(page::flags::queued);
+        running_.push_back(page);
+
+        if (page.test_flag(page::flags::read)) {
+            dispatch_read(page, std::move(units));
+
+        } else if (page.test_flag(page::flags::write)) {
+            dispatch_write(page, std::move(units));
+
+        } else {
+            assert(false);
+            std::terminate();
+        }
+    }
+}
+
+void io_queue::dispatch_read(
+  page& page, seastar::semaphore_units<> units) noexcept {
+    log.debug("Reading {} at {}~{}", path_, page.offset(), page.data().size());
+
+    std::ignore = seastar::with_gate(
+      gate_, [this, &page, units = std::move(units)]() mutable {
+          return file_
+            ->dma_read(
+              page.offset(), page.data().get_write(), page.data().size())
+            .then([this, &page, units = std::move(units)](auto) {
+                running_.erase(request_list_type::s_iterator_to(page));
+                if (complete_) {
+                    complete_(page);
+                }
+            })
+            .handle_exception([this](std::exception_ptr eptr) {
+                log.error("Read failed to {}: {}", path_, eptr);
+                std::terminate();
+            });
+      });
+}
+
+void io_queue::dispatch_write(
+  page& page, seastar::semaphore_units<> units) noexcept {
+    log.debug("Writing {} at {}~{}", path_, page.offset(), page.data().size());
+
+    std::ignore = seastar::with_gate(
+      gate_, [this, &page, units = std::move(units)]() mutable {
+          return file_
+            ->dma_write(page.offset(), page.data().get(), page.data().size())
+            .then([this, &page, units = std::move(units)](auto) {
+                running_.erase(request_list_type::s_iterator_to(page));
+                if (page.test_flag(page::flags::queued)) {
+                    pending_.push_back(page);
+                    cond_.signal();
+                }
+                if (complete_) {
+                    complete_(page);
+                }
+            })
+            .handle_exception([this](std::exception_ptr eptr) {
+                log.error("Write failed to {}: {}", path_, eptr);
```

Comment:
I think this log message is misleading because the `complete_` callback may also throw.

Reply:
good catch. i made the callback signature require noexcept.